http-gateways/IPLD_GATEWAY.md

```diff
@@ -0,0 +1,196 @@
+
+The path segments will be traversed with any parameters used to transform data along the way.
+
+The `format` query string parameter, or the `Accept` request header can be used to control the format which will be used to return the data.
+
+By default, data will be returned as `dag-json`
+
+<!--
+TODO: Cache control semantics?
+TODO: Add more details on how the traversal works?
+-->
+
+### `HEAD /ipld/{cid}[/{segments}][?{params}]`
+
+Resolves IPLD paths, and yields the same status code and headers as `GET`.
+
+### `POST /ipld/` and `POST /ipld/localhost/`
+
+Upload raw data to IPLD.
+The `body` of the request shall be parsed according to the `Content-Type` as IPLD data via standard encodings.
+`/localhost/` is used to support `POST ipld://localhost/` for uploading IPLD data to local nodes in web browsers that support it.
+
+The response will contain an `ipld://{cid}/` URL pointing at your data.
+
+<!--
+TODO: Only allow `/localhost/`? Get rid of `/localhost` from the spec if light clients with protocol handlers don't matter/
+-->
+
+### `PATCH /ipld/{cid}[/{segments}][?{params}]`
+
+This endpoint enables you to apply a [patch set](https://ipld.io/specs/patch/) to existing IPLD data.
+
+<!--
+TODO: Talk about content encoding
+TODO: Mention interaction with ADLs/Schemeas/Selectors
+-->
+
+The response will be an `ipld://` URL with your updated data.
+
+Note that the CID in the response URL will contain the same `segments` as in the request URL.
+e.g. if you patch data at `ipld://{cid1}/some/path/`, you will get back a URL that looks like `ipld://{updated cid}/some/path/`
+This enables you to make complex changes to a subtree in a dataset and get back a new root CID to use in your application.
+
+## HTTP Request
+
+### Request Headers
+
+#### `Accept` (request header)
+
+For `/ipld/{cid}/*` paths, the `Accept` header is used to indicate the encoding that should be used to return the data.
+This means that data initially encoded as `dag-json` will be transcoded to `dag-cbor` if the `application/vnd.ipld.dag-cbor` Accept header is used.
+
+- `application/json`: Interpret in the same way as `application/vnd.ipld.dag-json`.
+- `application/vnd.ipld.dag-json`: Return the block specified by the path encoded in `dag-json`.
+- `application/vnd.ipld.dag-cbor`: Return the block specified by the path ecnoded in `dag-cbor`.
+
+If no `Accept` header is present in the request, it will be assumed to be `application/vnd.ipld.dag-json`.
+
+#### `Content-Type` (request header)
+
+This header applies to `PUT/POST/PATCH` requests for `/ipld/*` paths on IPLD gateways which also support the [writable gateways spec](./WRITABLE_GATEWAY.md).
+Including it will hint to the writable IPLD gateway which encoding to use to parse the request body into the IPLD Data Model.
+
+- `application/json`: Interpret in the same way as `application/vnd.ipld.dag-json`.
+- `application/vnd.ipld.dag-json`: Return the block specified by the path encoded in `dag-json`.
```

Comment:
TODO: link to IANA when https://github.com/ipfs/in-web-browsers/issues/202 is resolved